### PR TITLE
JDK-8312415: Expand -Xlint:serial checks to enum constants with specialized class bodies

### PR DESCRIPTION
**Pull request by jddarcy**
Thanks to @lahodaj for providing the code to bridge into the class used for a enum constant with a specialized body.

This changes expands serialization warnings to the contents of any such class bodies.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8312415](https://bugs.openjdk.org/browse/JDK-8312415): Expand -Xlint:serial checks to enum constants with specialized class bodies (**Enhancement** - P4)

### Reviewers
 * [Jonathan Gibbons](https://openjdk.org/census#jjg) (@jonathan-gibbons - **Reviewer**) ⚠️ Review applies to [d61f9114](https://git.openjdk.org/jdk/pull/15004/files/d61f9114ab7df59c32fa0e8e6e53007c6d746139)
 * [Jan Lahoda](https://openjdk.org/census#jlahoda) (@lahodaj - **Reviewer**) ⚠️ Review applies to [d61f9114](https://git.openjdk.org/jdk/pull/15004/files/d61f9114ab7df59c32fa0e8e6e53007c6d746139)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15004/head:pull/15004` \
`$ git checkout pull/15004`

Update a local copy of the PR: \
`$ git checkout pull/15004` \
`$ git pull https://git.openjdk.org/jdk.git pull/15004/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15004`

View PR using the GUI difftool: \
`$ git pr show -t 15004`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15004.diff">https://git.openjdk.org/jdk/pull/15004.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15004#issuecomment-1648499942)

### CODE REVIEW

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Check.java [hunk 1]
```diff
@@ -5213,6 +5213,18 @@ public Void visitTypeAsEnum(TypeElement e,
                             }
                         }
                     }
+
+                    case ENUM_CONSTANT -> {
+                        var field = (VarSymbol)enclosed;
+                        if (field.isEnum()) {
```
Comment (jonathan-gibbons) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Check.java:5219]: It's not wrong, but is this check redundant?  When can an `ENUM_CONSTANT` not be `isEnum()` ?

**Review by jonathan-gibbons — APPROVED** [review]
Approved, but see question for Check.java

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Check.java (hunk 1 above)
Comment (lahodaj) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Check.java:5219]: Yes, sorry for that, I've originally had the code elsewhere. The `isEnum` check should not be needed here.

**Review by lahodaj — APPROVED** [review]
Looks reasonable to me.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Check.java (hunk 1 above)
Comment (jddarcy) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Check.java]: Code updated; all langtools tests pas when run locally. Thanks.

**Pull request closed without merging**